pkg/driver/node/mounter/pod_mounter.go

```diff
@@ -0,0 +1,403 @@
+
+// mountSyscall is the function that performs `mount` operation for given `target` with given Mountpoint `args`.
+// It returns mounted FUSE file descriptor as a result.
+// This is mainly exposed for testing, in production platform-native function (`mountSyscallDefault`) will be used.
+type mountSyscall func(target string, args mountpoint.Args) (fd int, err error)
+
+// A PodMounter is a [Mounter] that mounts Mountpoint on pre-created Kubernetes Pod running in the same node.
+type PodMounter struct {
+	client                 k8sv1.CoreV1Interface
+	mountpointPodNamespace string
+	mount                  mount.Interface
+	kubeletPath            string
+	mountSyscall           mountSyscall
+	kubernetesVersion      string
+	credProvider           *credentialprovider.Provider
+}
+
+// NewPodMounter creates a new [PodMounter] with given Kubernetes client.
+func NewPodMounter(client k8sv1.CoreV1Interface, credProvider *credentialprovider.Provider, mountpointPodNamespace string, mount mount.Interface, mountSyscall mountSyscall, kubernetesVersion string) (*PodMounter, error) {
+	return &PodMounter{
+		client:                 client,
+		mountpointPodNamespace: mountpointPodNamespace,
+		mount:                  mount,
+		kubeletPath:            util.KubeletPath(),
+		mountSyscall:           mountSyscall,
+		kubernetesVersion:      kubernetesVersion,
+		credProvider:           credProvider,
+	}, nil
+}
+
+// Mount mounts the given `bucketName` at the `target` path using provided credential context and Mountpoint arguments.
+//
+// At high level, this method will:
+//  1. Wait for Mountpoint Pod to be `Running`
+//  2. Write credentials to Mountpoint Pod's credentials directory
+//  3. Obtain a FUSE file descriptor
+//  4. Call `mount` syscall with `target` and obtained FUSE file descriptor
+//  5. Send mount options (including FUSE file descriptor) to Mountpoint Pod
+//  6. Wait until Mountpoint successfully mounts at `target`
+//
+// If Mountpoint is already mounted at `target`, it will return early at step 2 to ensure credentials are up-to-date.
+func (pm *PodMounter) Mount(ctx context.Context, bucketName string, target string, credentialCtx credentialprovider.ProvideContext, args mountpoint.Args) error {
+	podID, volumeID := credentialCtx.PodID, credentialCtx.VolumeID
+
+	err := pm.checkTargetPath(target)
+	if err != nil {
+		return fmt.Errorf("Failed to check target path %q: %w", target, err)
+	}
+
+	isMountPoint, err := pm.IsMountPoint(target)
+	if err != nil {
+		return fmt.Errorf("Could not check if %q is a mount point: %w", target, err)
+	}
+
+	// TODO: If `target` is a `systemd`-mounted Mountpoint, this would return an error,
+	// but we should still update the credentials for it by calling `credProvider.Provide`.
+	pod, podPath, err := pm.waitForMountpointPod(ctx, podID, volumeID)
+	if err != nil {
+		return fmt.Errorf("Failed to wait for Mountpoint Pod for %q: %w", target, err)
```

Comment:
"wait for Mountpoint pod to be ready for %q"

Reply:
Also, can we include the Mountpoint pod name/namespace here if possible?

Reply:
Here the `pod` might be nil, but the function `waitForMountpointPod` logs Mountpoint Pod name in the cases where it's available

Reply:
First part addressed with https://github.com/awslabs/mountpoint-s3-csi-driver/pull/368/commits/ea593719beebcf7db4d3f8d83f34710e2da74e38